src/services/factories/tokenList.ts

```diff
@@ -130,42 +130,57 @@ export function getTokensFactory(factoryParams: {
+    if (tcrAddressesSet.size > 0) {
+      // If filtering by TCR
+      logDebug(`[tokenListFactory][${networkId}] TCR contains ${tcrAddressesSet.size} addresses`)
+      logDebug(`[tokenListFactory][${networkId}] Token id and address mapping:`)
+      const filteredAddressAndIds = Array.from(listedAddressesAndIds.entries()).filter(([address, id]) => {
+        // Remove addresses that are not on the TCR, if any
+        const isInTcr = tcrAddressesSet.has(address)
+        logDebug(`[tokenListFactory][${networkId}] ${id} : ${address}. On TCR? ${isInTcr}`)
+      })
+      return new Map<string, number>(filteredAddressAndIds)
+    } else {
+      // If not using a TCR, filter by default list
+      logDebug(`[tokenListFactory][${networkId}] Not using a TCR. Filtering by tokens in the config file`)
+      const tokensConfig = tokenListApi.getTokens(networkId)
+      const filteredAddressAndIds = tokensConfig
+        .map((token): [string, number] | null => {
+          const tokenId = listedAddressesAndIds[token.address]
+          return [token.address, tokenId] || null
+        })
+        .filter(notEmpty)
+
+      return new Map<string, number>(filteredAddressAndIds)
+    }
+  }
 
-    logDebug(`[tokenListFactory][${networkId}] Token id and address mapping:`)
-    addressesAndIds.forEach((id, address) => {
-      // Remove addresses that are not on the TCR, if any
-      let removed = false
-      if (tcrAddressesSet.size > 0 && !tcrAddressesSet.has(address)) {
-        addressesAndIds.delete(address)
-        removed = true
-      }
-      logDebug(`[tokenListFactory][${networkId}] ${id} : ${address}. On TCR? ${!removed}`)
-    })
+  async function updateTokenDetails(networkId: number, numTokens: number): Promise<void> {
+    // Get filtered ids and addresses
+    const filteredAddressesAndIds = await _getFilteredIdsMap(networkId, numTokens)
 
     // Create a map of current token list addresses and tokens
     const localAddressesMap = new Map<string, TokenDetails>(
       tokenListApi
-        // Get a copy of the current token list
         .getTokens(networkId)
         // Remove tokens that are on the list but not registered on the contract neither on TCR
-        .filter(({ address }) => addressesAndIds.has(address))
+        .filter(({ address }) => filteredAddressesAndIds.has(address))
         // Map it with token address as key for easy access
         .map((token) => [token.address, token]),
     )
 
     const promises: Promise<TokenErc20 | string>[] = []
 
     // Go over all value (id) and key (tokenAddress) pairs registered on the contract
-    addressesAndIds.forEach((id, tokenAddress) => {
+    filteredAddressesAndIds.forEach((id, tokenAddress) => {
       if (!localAddressesMap.has(tokenAddress)) {
         // New token not in our local list, fetch erc20 details for it
         logDebug(`[tokenListFactory][${networkId}] Address ${tokenAddress} with id ${id} not in local list, fetching`)
```

Comment:
This looks complicated

Is the idea to:

1. get intersection of tokens from `tokenListApi.getTokens(networkId)` and `filteredAddressesAndIds`
2. fetch tokens that are in `filteredAddressesAndIds` but not in `tokenListApi.getTokens(networkId)` as ERC20
3. fill in the blanks of tokens from 2 from chain as ERC20
4. assign ids to fetched ERC20

If I'm correct, then I feel you have too many loops and a pretty complicated logic in general. Would the following do the same?

```ts
async function updateTokenDetails(networkId: number, numTokens: number): Promise<void> {
  // Get filtered ids and addresses
  const filteredAddressesAndIds = await _getFilteredIdsMap(networkId, numTokens)
  const localTokens: TokenDetails[] = []

  const allTokensFromApi = new Map(tokenListApi.getTokens(networkId).map((t) => [t.address, t]))
  const TokenDetailsPromises: Promise<TokenDetails | undefined>[] = []

  filteredAddressesAndIds.forEach((id, tokenAddress) => {
    const token = allTokensFromApi.get(tokenAddress)
    if (token) {
      localTokens.push(token)
    } else {
      const fullTokenPromise = getErc20DetailsOrAddress(networkId, tokenAddress).then((partialToken) => {
        if (typeof partialToken === 'string') {
          // We replaced potential null responses with original tokenAddress string for logging purposes
          logDebug(`[tokenListFactory][${networkId}] Address ${partialToken} is not a valid ERC20 token`)
          return
        }
        // If we got a valid response
        logDebug(
          `[tokenListFactory][${networkId}] Got details for address ${partialToken.address}: symbol '${partialToken.symbol}' name '${partialToken.name}'`,
        )
        // build token object
        const token: TokenDetails = { ...partialToken, id }

        return token
      })
      TokenDetailsPromises.push(fullTokenPromise)
    }
  })

  const fetchedAndFilledTokenDetails = (await Promise.all(TokenDetailsPromises)).filter(notEmpty)

  const tokenList = localTokens.concat(fetchedAndFilledTokenDetails)

  // Persist it \o/
  tokenListApi.persistTokens({ networkId, tokenList })
}
```

Reply:
@anxolin did you consider this refactor?

Reply:
:point_up: @Velenir's logic looks correct and is easier to read imo

Reply:
Proposal that reiterate on Dima's ideas. I think it simplifies it a bit further and fix some weird signature we had in the API
https://github.com/gnosis/dex-react/pull/1498/files